use-shopping-cart/react/index.js

```diff
@@ -0,0 +1,92 @@
+import * as React from 'react'
+import { actions, cartInitialState } from '../core/slice'
+import { createShoppingCartStore } from '../core/store'
+import { createDispatchHook, createSelectorHook, Provider } from 'react-redux'
+import { bindActionCreators } from '@reduxjs/toolkit'
+import { filterCart } from '../utilities/old-utils'
+//TODO figure out how to apply formatCurrencyString
+import { formatCurrencyString } from '../core/store'
+import { PersistGate } from 'redux-persist/integration/react'
+import { persistStore } from 'redux-persist'
+
+export { actions, filterCart, formatCurrencyString }
+export const CartContext = React.createContext(cartInitialState)
+export const useSelector = createSelectorHook(CartContext)
+export const useDispatch = createDispatchHook(CartContext)
+
+export function CartProvider({ children, ...props }) {
+  const store = React.useMemo(() => createShoppingCartStore(props), [props])
+  const persistor = persistStore(store)
+
+  return (
+    <Provider context={CartContext} store={store}>
+      <PersistGate
+        persistor={persistor}
+        children={(bootstrapped) => {
+          if (!bootstrapped) {
+            return <h1>Loading . . . </h1>
+          }
+
+          return children
+        }}
+      />
```

Comment:
@dayhaysoos I think this is the wrong way to be using `<PersistGate>` in a library.

1. We shouldn't handle loading UI for the developers as they will want their own custom loading screen.
2. This loading UI you've provided by default is inaccessible.
3. The redux-persist docs provide a very easy to use `loading` prop like so:

```jsx
<PersistGate loading={<Loading />} persistor={persistor} />
```

With the `loading` prop, we can just allow the developer to pass down their own `loading` prop in place of it and we can default to `null`.

```js
function CartProvider({ loading = null, children, ...props }) {
  return (
    <Provider {/* ... */}>
      <PersistGate {/* ... */} loading={loading} children={children} />
    </Provider>
  )
}
```

Reply:
Oops! Yeah I didn't intend to leave that, this was a result of me troubleshooting

Reply:
So I just tried using PersistGate as intended from the docs and I don't think it's possible. All tests break. Feels like the only way to go about this would be sticking to the `children` function. I passed the `loading` prop to be returned if bootstrapped is `false`

Reply:
Can I see what you tried for the `<PersistGate>`? I'm not sure you'd ever have a `bootstrapped` if you used the `loading` prop.

Reply:
@ChrisBrownie55 I did exactly what you have above, that's how I initially tried to do it. Persist Gate refers to `bootstrapped` under the hood the same way, bootstrapped initially is false and it displays what was passed to `loading` in the meantime. Once `bootstrapped` turns to `true` it renders `children`. When I use the simple way you see in the docs, `bootstrapped` stays false the entire time. For some reason doing the children's function is the only thing I've managed to get working. I don't quite understand why this is the case. I feel like we should make a separate branch to troubleshoot this, but as of right now, there's no problem with the way this is being done now that we are allowing them to pass a loading prop.

Reply:
@dayhaysoos It seems to work both ways. I make a CodeSandbox that shows that it does work with the `loading` prop.
https://codesandbox.io/s/redux-persist-demo-ogu5v?file=/src/index.js

I added an artificial delay to demonstrate that the loading screen will in fact show. If the tests are failing, I would think it's because they were written for the way React handled the `useLocalStorageReducer` we had.


https://user-images.githubusercontent.com/19195374/111063319-d8ea2700-847b-11eb-9ee2-fa2af5cc2c23.mp4

Reply:
The tests weren't failing, it just wasn't rendering at all for Create React App. Would you be okay with leaving it the way it is for now and trying to figure out how to get the simpler implementation going later?

Reply:
Yeah, that'd be fine. If it works and the developer can still specify what it should look like while loading via a `loading` prop then it's good.